### ISSUE TITLE
Search with distinctAttribute returns offset+limit hits

### ISSUE THREAD

**Issue by mitallast** [meilisearch/meilisearch#4078]
Describe the bug
Search with distinctAttribute and non-zero offset returns `offset+limit` hits, instead of `limit`. 

To Reproduce
Make an index with `distinctAttribute`,  without `ranking_rules`. 
Add 40 documents with 20 distinct attribute values.
Empty search request, only limit=10, offset=20.

Expected behavior
10 hits, same as `limit`

Actual behavioir
30 hits, actually `offset+limit`

Screenshots

Meilisearch version:
1.3.5

Additional context
Looks like a bug in bucket_sort.rs - `results` contains matches max size `limit+offset`, and there's no drop first `offset` matches. 

https://github.com/meilisearch/meilisearch/blob/bb79bdb3f8e04afdc531ea663b93c47317b118c9/milli/src/search/new/bucket_sort.rs#L48-L65

**Comment by curquiza** [comment]
Hello @mitallast 

> Make an index with `distinctAttribute`, without `ranking_rules`.

Not sure what you mean with this

Can you give more details about the steps you make to describe the issue you have? We need to be able to reproduce it in order to fix it

To simplify it for you, you can maybe give an example with only 10 hits?

**Comment by mitallast** [comment]
I apologize for the long example, I shortened it as much as possible:

1) create index

```http
POST localhost:7700/indexes
Content-Type: application/json

{
  "uid": "test",
  "primaryKey": "productId"
}
```

2) set up distinctAttribute

```http
PATCH localhost:7700/indexes/test/settings
Content-Type: application/json

{
  "distinctAttribute": "shopId"
}
```

3) add 10 documents

```http
POST localhost:7700/indexes/test/documents
Content-Type: application/json

[
  {"productId": 1, "shopId": 1},
  {"productId": 2, "shopId": 1},
  {"productId": 3, "shopId": 2},
  {"productId": 4, "shopId": 2},
  {"productId": 5, "shopId": 3},
  {"productId": 6, "shopId": 3},
  {"productId": 7, "shopId": 4},
  {"productId": 8, "shopId": 4},
  {"productId": 9, "shopId": 5},
  {"productId": 10, "shopId": 5}
]
```

4) try search first 2 documents, expected 2 hits

```http
GET localhost:7700/indexes/test/search?limit=2
```

response has 2 hits as expected, same as limit:

```http
HTTP/1.1 200 OK
content-type: application/json

{
  "hits": [
    {"productId": 1, "shopId": 1},
    {"productId": 4, "shopId": 2}
  ],
  "query": "",
  "processingTimeMs": 0,
  "limit": 2,
  "offset": 0,
  "estimatedTotalHits": 8
}
```

5) try search next 2 documents with offset 2, expected 2 hits

```http
GET localhost:7700/indexes/test/search?limit=2&offset=2
```

response has 4 hits instead of 2, like limit+offset=2+2=4, the first two hits match the previous request:

```http
HTTP/1.1 200 OK
content-type: application/json

{
  "hits": [
    {"productId": 1, "shopId": 1},
    {"productId": 4, "shopId": 2},
    {"productId": 5, "shopId": 3},
    {"productId": 7, "shopId": 4}
  ],
  "query": "",
  "processingTimeMs": 0,
  "limit": 2,
  "offset": 2,
  "estimatedTotalHits": 6
}
```

6) try search with offset 10, empty response expected:

```http
GET localhost:7700/indexes/test/search?limit=2&offset=10
```

response has 5 hits, instead of empty result

```http
HTTP/1.1 200 OK
content-type: application/json

{
  "hits": [
    {"productId": 1, "shopId": 1},
    {"productId": 4, "shopId": 2},
    {"productId": 5, "shopId": 3},
    {"productId": 7, "shopId": 4},
    {"productId": 9, "shopId": 5}
  ],
  "query": "",
  "processingTimeMs": 0,
  "limit": 2,
  "offset": 10,
  "estimatedTotalHits": 5
}
```

**Comment by curquiza** [comment]
Hello @mitallast 
Indeed, I succeeded in reproducing the issue with your examples. Thanks for the really clear examples!
I get more hits than required!

I discussed with the developers of the @meilisearch/engine-team  and I come back to you!

**Comment by curquiza** [comment]
More investigation: this bug has been introduced in v1.2.0

**Comment by montassar-vaionex** [comment]
same issue in "meilisearch": "^0.35.0",

**Comment by curquiza** [comment]
Hello again people following this issue

The team cannot spend time fixing the bug now, but we are available to review any PR fixing it.

@mitallast you have indeed spotted the part that needs to be changed in the code.

For the contribution, we don't recommend contributing if you have never contributed in Rust, because the code bsae is auite big.

⚠ We need to add tests to ensure this bug will not happen again

**Comment by vivek-26** [comment]
Opened PR #4108 to fix this issue. Requesting review.
cc/ @curquiza @Kerollmops

**Comment by curquiza** [comment]
Thank you @vivek-26
@Kerollmops or @ManyTheFish will review it when they have some time! 😊

**Comment by curquiza** [comment]
Hello @bultas 

We are supposed to release it tomorrow.
However we have an issue with [a PR](https://github.com/meilisearch/meilisearch/pull/4089) merging currently. And this PR must absolutely be in v1.4.1
So if we cannot make it tomorrow, it will be beginning of next week (Monday or Tuesday maximum)
If the issue is more complex than we expect, I'll let you know

**Comment by bultas** [comment]
@curquiza thanks for the quick info

**Comment by curquiza** [comment]
Fixed by #4108 

Launching CIs for the release in the coming hour since all PRs for v1.4.1 have been merged! 🎉
